### PR TITLE
keys: Add a rotate-targets command

### PR DESCRIPTION
**Pull request by doanac**
This allows a user to have their own private key capable of signing
targets.json

Signed-off-by: Andy Doan <andy@foundries.io>

### CODE REVIEW

**Comment by doanac** [comment]
This is really a single commit based on:

 https://github.com/foundriesio/fioctl/pull/83

It requires a new backend resource - https://git.foundries.io/development/cloudplatforms/infrastructure/ota-lite/-/merge_requests/104

File: client/foundries.go [hunk 1]
```diff
@@ -862,6 +862,17 @@ func (a *Api) FactoryListDeviceGroup(factory string) (*[]DeviceGroup, error) {
 	return &resp.Groups, nil
 }
 
+func (a *Api) CiTargets(factory string) (*AtsKey, error) {
```
Comment (vkhoroz) [inline client/foundries.go:865]: Can you rename this into smth like `GetOnlineTargetsKey` ?
As is I constantly have a feeling that you fetch targets when calling this method.

File: subcommands/keys/rotate_targets.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+	block = &pem.Block{
+		Type:  "PUBLIC KEY",
+		Bytes: pubBytes,
+	}
+	pubBytes = pem.EncodeToMemory(block)
+	id := fmt.Sprintf("%x", sha256.Sum256(pubBytes))
+	root.Signed.Keys[id] = client.AtsKey{
+		KeyType:  "RSA",
+		KeyValue: client.AtsKeyVal{Public: string(pubBytes)},
+	}
```
Comment (vkhoroz) [inline subcommands/keys/rotate_targets.go:117]: Around 20 lines from above are common with root key rotation;  and I already see some tiny differences (well, discrepancies).  I think they deserve to be extracted into a separate function.

File: client/foundries.go (hunk 1 above)
Comment (doanac) [reply client/foundries.go]: https://github.com/foundriesio/fioctl/pull/84/commits/9b5c948da0ef274d59c51c22bc3173f21bf18046

File: subcommands/keys/rotate_targets.go (hunk 2 above)
Comment (doanac) [reply subcommands/keys/rotate_targets.go]: not that pretty, but consolidated - https://github.com/foundriesio/fioctl/pull/84/commits/14e292e1b25fbf6a44b4db98b4ffac4e012e5a3b

once reviewed, i'll squash that into the prior commit.

File: subcommands/keys/rotate_targets.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	for _, keyid := range root.Signed.Roles["targets"].KeyIDs {
+		pub := root.Signed.Keys[keyid].KeyValue.Public
+		if pub == onlinePub.KeyValue.Public {
+			return keyid, nil
+		}
+	}
+	return "", errors.New("Unable to find online target key for factory")
+}
+
+func replaceProdTargetKey(root *client.AtsTufRoot, onlineTargetId string, creds OfflineCreds) (string, OfflineCreds) {
```
Comment (vkhoroz) [inline subcommands/keys/rotate_targets.go:84]: Can you replace Prod with Offline everywhere ?
This patch set doesn't even create prod root role - it's CI root role, which has online and offline targets keys.

I'm scary we might overload a word "prod" rather quickly if we put it in every place which has something to do with production environment.

**Review by vkhoroz — APPROVED** [review]

File: subcommands/keys/rotate_targets.go (hunk 2 above)
Comment (vkhoroz) [reply subcommands/keys/rotate_targets.go]: I like how it looks, even better than I thought ;-)

**Comment by vkhoroz** [comment]
LGTM

File: subcommands/keys/rotate_targets.go (hunk 3 above)
Comment (doanac) [reply subcommands/keys/rotate_targets.go]: good point: https://github.com/foundriesio/fioctl/pull/84/commits/26c454a725366fb61d10a83d015f7d95d6966a86

if that looks good, i'll squash fix ups and merge

Comment (vkhoroz) [reply subcommands/keys/rotate_targets.go]: looks good

**Comment by doanac** [comment]
squash fixed everything.

**Pull request merged**